src/common/reduction/clustering/Hierarchical2DClusterer.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) 2023 European Spallation Source, ERIC. See LICENSE file
+//===----------------------------------------------------------------------===//
+///
+/// \file Hierarchical2DClusterer.cpp
+/// \brief Hierarchical2DClusterer class implementation
+///
+//===----------------------------------------------------------------------===//
+
+#include <common/debug/Trace.h>
+#include <common/reduction/clustering/Hierarchical2DClusterer.h>
+#include <set>
+// #undef TRC_LEVEL
+// #define TRC_LEVEL TRC_L_DEB
+
+Hierarchical2DClusterer::Hierarchical2DClusterer(uint64_t max_time_gap,
+                                                 uint16_t max_coord_gap)
+    : Abstract2DClusterer(), max_time_gap_(max_time_gap), 
+    // Store the sqr of the max coord gap to reduce computation during clustering
+    max_coord_gap_(max_coord_gap), max_coord_gap_sqr_(sqr(max_coord_gap)) {}
+
+void Hierarchical2DClusterer::insert(const Hit2D &hit) {
+  /// Process time-cluster if time gap to next hit is large enough
+  if (!current_time_cluster_.empty() &&
+      (hit.time - current_time_cluster_.back().time) > max_time_gap_) {
+    flush();
+  }
+
+  /// Insert hit in either case
+  current_time_cluster_.emplace_back(hit);
+}
+
+void Hierarchical2DClusterer::cluster(const Hit2DVector &hits) {
+  /// It is assumed that hits are sorted in time
+  for (const auto &hit : hits) {
+    insert(hit);
+  }
+}
+
+void Hierarchical2DClusterer::flush() {
+  XTRACE(EVENT, DEB, "Flushing clusterer");
+  if (current_time_cluster_.empty()) {
+    return;
+  }
+  cluster_by_x();
+  current_time_cluster_.clear();
+}
+
+void Hierarchical2DClusterer::cluster_by_x() {
+  uint clusterSize = current_time_cluster_.size();
+  std::vector<bool> visited(clusterSize, false); // keep track of visited points
+
+  XTRACE(DATA, DEB, "%u events in time window", clusterSize);
+  for (uint i = 0; i < clusterSize; i++) {
+    if (visited[i]) {
+      continue; // skip points that have already been visited
+    }
+    XTRACE(DATA, DEB, "Starting new cluster");
+    Hit2DVector space_cluster; // initialize a new cluster
+    space_cluster.push_back(current_time_cluster_[i]);
+    visited[i] = true;
+    for (uint j = i + 1; j < clusterSize; j++) {
+      if (visited[j]) {
+        continue; // skip points that have already been visited
+      }
+      double x_distance = (double)current_time_cluster_[i].x_coordinate -
+                          (double)current_time_cluster_[j].x_coordinate;
+      double y_distance = (double)current_time_cluster_[i].y_coordinate -
+                          (double)current_time_cluster_[j].y_coordinate;
+
+      // Calculate distance according to d2 = dx2 + dy2 to remove sqrt calculation
+      double distance_sqr = sqr(x_distance) + sqr(y_distance);
```

Comment:
Sorry for the spam, but I realized that I made a typo, so I corrected it. 
@mortenjc could you have a look on this change. I added an inline sqr function and also compare sqr(threshold) with sqr(distance). Now pow() and sqrt() eliminated.